### PR TITLE
[SWDEV-396381] FSDP UT failures

### PR DESCRIPTION
**Pull request by lcskrishna**
This PR fixes UTs:  test_fsdp_uneven and test_fsdp_tp_integration in SWDEV-396381. 
This PR limits the nranks to 8GPUs if there are more than 8GPUs in a node. This avoids any key error trying to access for nodes greater than 8 as the test seems to be very small.

### CODE REVIEW

**Comment by pruthvistony** [comment]
We have a file CHANGELOG.md on rocm5.7_internal_testing branch which should be updated when any feature/UTs are enabled. It is NOT yet followed properly, but its required to start someday. So request you to update the CHANGELOG.md file with a one line description of the change and add to this PR.

File: torch/testing/_internal/common_fsdp.py [hunk 1]
```diff
@@ -823,7 +823,12 @@ def setUp(self):
 
     @property
     def world_size(self):
-        return torch.cuda.device_count() if torch.cuda.is_available() else 4
+        nranks = 4
+        if torch.cuda.is_available:
+            nranks = torch.cuda.device_count()
+        if torch.version.hip and nranks > 8:
```
Comment (jithunnair-amd) [inline torch/testing/_internal/common_fsdp.py:829]: Why should this change be HIP-specific?

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]
@lcskrishna Can you please elaborate more on the root-cause here? I couldn't glean from your description - "This avoids any key error trying to access for nodes greater than 8 as the test seems to be very small." - whether this change impacts only ROCm. It sounds like it's more about the number of GPUs, but some more explanation about the "key error" you refer to would be helpful.

File: torch/testing/_internal/common_fsdp.py (hunk 1 above)
Comment (lcskrishna) [reply torch/testing/_internal/common_fsdp.py:829]: @jithunnair-amd you are right, this is applicable to CUDA as well as long as there is a node with more than 8 GPUs.

**Comment by lcskrishna** [comment]
> @lcskrishna Can you please elaborate more on the root-cause here? I couldn't glean from your description - "This avoids any key error trying to access for nodes greater than 8 as the test seems to be very small." - whether this change impacts only ROCm. It sounds like it's more about the number of GPUs, but some more explanation about the "key error" you refer to would be helpful.

There are few unit tests in FSDP that can support upto 8 GPUs. 
In this case, for example test_fsdp_uneven has an input size of [8,3] and which infact for each process/rank we pass the data as input[self.rank] as below. So when we use 16 GPUs for our tests, these tests throw an index/key error. So basically to avoid such corner cases, I would like to add this change to use 8GPUs if there are more than 8 GPUs. This is applicable to CUDA builds as well. 

https://github.com/pytorch/pytorch/blob/main/test/distributed/fsdp/test_fsdp_uneven.py#L44
https://github.com/pytorch/pytorch/blob/main/test/distributed/fsdp/test_fsdp_uneven.py#L55

File: torch/testing/_internal/common_fsdp.py (hunk 1 above)
Comment (jithunnair-amd) [reply torch/testing/_internal/common_fsdp.py:829]: In that case, I'd suggest the following patch:
```
- return torch.cuda.device_count() if torch.cuda.is_available() else 4
+ return min(torch.cuda.device_count(), 8) if torch.cuda.is_available() else 4
```
And let's file an upstream PR with the same patch, with the description containing what you explained in https://github.com/ROCmSoftwarePlatform/pytorch/pull/1234#issuecomment-1580213588

**Review by jithunnair-amd — APPROVED** [review]

**Review by jithunnair-amd — APPROVED** [review]
